examples/custom_shader/custom_pass.rs

```diff
@@ -0,0 +1,348 @@
+///         ShaderKind::Vertex,
+///         SourceLanguage::GLSL,
+///         "main",
+///     ).precompile().unwrap();
+///
+///     static ref VERTEX: SpirvShader = SpirvShader::new(
+///         (*VERTEX_SRC).spirv().unwrap().to_vec(),
+///         (*VERTEX_SRC).stage(),
+///         "main",
+///     );
+/// }
+/// '''
+
+/// Draw triangles.
+#[derive(Clone, Debug, PartialEq, Derivative)]
+#[derivative(Default(bound = ""))]
+pub struct DrawCustomDesc;
+
+impl DrawCustomDesc {
+    /// Create instance of `DrawCustomDesc` render group
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl<B: Backend> RenderGroupDesc<B, World> for DrawCustomDesc {
+    fn build(
+        self,
+        _ctx: &GraphContext<B>,
+        factory: &mut Factory<B>,
+        _queue: QueueId,
+        _world: &World,
+        framebuffer_width: u32,
+        framebuffer_height: u32,
+        subpass: hal::pass::Subpass<'_, B>,
+        _buffers: Vec<NodeBuffer>,
+        _images: Vec<NodeImage>,
+    ) -> Result<Box<dyn RenderGroup<B, World>>, failure::Error> {
+        let env = DynamicUniform::new(factory, pso::ShaderStageFlags::VERTEX)?;
+        let vertex = DynamicVertexBuffer::new();
+
+        let (pipeline, pipeline_layout) = build_custom_pipeline(
+            factory,
+            subpass,
+            framebuffer_width,
+            framebuffer_height,
+            vec![env.raw_layout()],
+        )?;
+
+        Ok(Box::new(DrawCustom::<B> {
+            pipeline,
+            pipeline_layout,
+            env,
+            vertex,
+            vertex_count: 0,
+        }))
+    }
+}
+
+/// Draws triangles to the screen.
+#[derive(Debug)]
+pub struct DrawCustom<B: Backend> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    env: DynamicUniform<B, CustomUniformArgs>,
+    vertex: DynamicVertexBuffer<B, CustomArgs>,
+    vertex_count: usize,
+}
+
+impl<B: Backend> RenderGroup<B, World> for DrawCustom<B> {
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        let (triangles,) = <(ReadStorage<'_, Triangle>,)>::fetch(world);
+
+        // Get our scale value
+        let scale = world.read_resource::<CustomUniformArgs>();
+
+        // Write to our DynamicUniform
+        self.env.write(factory, index, scale.std140());
+
+        self.vertex_count = triangles.join().count() * 3;
+
+        // Create an iterator over the Triangle vertices
+        let vertex_data_iter = triangles.join().flat_map(|triangle| triangle.get_args());
+
+        // Write the vector to a Vertex buffer
+        self.vertex.write(
+            factory,
+            index,
+            self.vertex_count as u64,
+            Some(vertex_data_iter.collect::<Box<[CustomArgs]>>()),
+        );
+
+        // Return that we request to draw
+        PrepareResult::DrawRecord
```

Comment:
You don't really need to record every frame with this. You need to record only when `self.vertex_count` have changed. Use `util::ChangeDetection` struct as we do in other passes to signal that correctly.

Reply:
That makes the assumption that the only changes require adding or removing a triangle. If a triangle is changed then it won't cause it to change. correct?

That being said all of this is not need for this example as it is static but i don't want to give an example that doesn't allow for updated entities.

Reply:
No, it will cause it to change. But changing the triangle data doesn't require a draw call. All you need to do is to update the buffer with new data.

You only need to redraw if any inputs to your commands have actually changed. This usually means just the number of objects have changed or a sequence of binds is different. The second never happens in this example, as your bind sequence is static.

Reply:
I will implement that change then.

Reply:
Implemented